### DISCUSSION TITLE
Correct usage of 'renderLinesAsTubes()'

### DISCUSSION THREAD

**Discussion by andama** [marcomusy/vedo#543] [Q&A] [answered]
Hi! Im trying to merge multiple lines and points into a mesh, then render the lines as tubes connecting the points instead of lines. Creating and merging the mesh is done the following way:

```python
import vedo as v
...

#Creates the mesh with only points at coordinates 'coord'
elements = v.Mesh([coord, None])

    for i in range(nel):
        # Coordinates for line ends are stored in an array, disregard this part
        coord1,coord2 = tools.get_coord_from_edof(edof[i,:],dof)
        dx = coord[coord2,0] - coord[coord1,0]
        dy = coord[coord2,1] - coord[coord1,1]
        dz = coord[coord2,2] - coord[coord1,2]
        x = coord[coord1,0] + 0.5*dx
        y = coord[coord1,1] + 0.5*dy
        z = coord[coord1,2] + 0.5*dz
            
        # Lines are created and merged with the existing mesh
        elements = v.merge([elements, v.Line(p0=[x-0.5*dx,y-0.5*dy,z-0.5*dz], p1=[x+0.5*dx,y+0.5*dy,z+0.5*dz])])
```

The mesh works and is rendered correctly appart from still being lines. This is despite renderLinesAsTubes() used on the mesh:

```python
self.vp += elements.renderLinesAsTubes(True)
```
Here 'self.vp' is the vedo plotter. I've also tried changing the global settings like this:
```python
v.settings.renderLinesAsTubes = True
```
Am I doing anything wrong? Is the 'renderLinesAsTubes()'-function applied incorrectly? I find it confusing that the global settings are not even working in that case, and suspect something else is wrong. I've tried using the function on the mesh after the merge is done, i.e. after the loop in the first code-snippet. Also tried using it when the mesh is created, both give the same result however.

Thanks in advance for any replies!

EDIT:
I've checked the 'vtkFollower' object, it returns the property:
```
RenderLinesAsTubes: On
```
So the function seems to be doing something, and yet the lines are not rendered as tubes.

**Comment by marcomusy** [comment]
Hi Andreas
maybe you just need to specify a lineWidth ? This seems to work:
```python
from vedo import *
s = Sphere(res=4)
s.lw(20).lc('orange8').renderLinesAsTubes()
show(s, axes=1)
```

<img width="424" alt="Screenshot 2021-11-22 at 15 07 12" src="https://user-images.githubusercontent.com/32848391/142876084-bd1ea1bc-13ce-4268-99b4-da8d164f70d3.png">

make sure you have the latest `vedo` version.

**Reply by andama** [reply]
Specifying the lineWidth works and they are (at least now) rendered as tubes. They don't behave like physical objects however which is what I need. I essentially need the lines to behave like cylinders and not scale the width when zooming like below. Is it working as intended? Do I need to use cylinders if I want them to not behave like this?

![Screenshot from 2021-11-22 15-16-45](https://user-images.githubusercontent.com/11843453/142877588-454bfc33-2c5f-485f-900d-e1442253b019.png)
![Screenshot from 2021-11-22 15-17-38](https://user-images.githubusercontent.com/11843453/142877594-ee374a9d-4506-4842-819c-8e846b35d4ef.png)

The issue is I want them as lines because if I need to use cylinders, it will create problems later when applying a 'cmap' to them. Version is 2021.0.6 as my repos haven't updated yet.

**Comment by marcomusy** [comment, marked as answer]
> Specifying the lineWidth works and they are (at least now) rendered as tubes. They don't behave like physical objects however which is what I need. I essentially need the lines to behave like cylinders and not scale the width when zooming like below. Is it working as intended? 

Yes, these are not "real" lines (in computer graphics are called "impostors" if i remember..), they simulate a tube but there is nothing physical there...

> Do I need to use cylinders if I want them to not behave like this?

Yes, I'm afraid... you can still create your tubes and associate a constant scalar id to all points/cells of these cylinders, so once you merge them you'll be able to use `cmap()`

**Reply by andama** [reply]
I understand, it seems I'd misunderstood what 'renderLinesAsTubes()' is intended for. I'll switch to cylinders. Merging works as intended, the problem is really with the scalar data I have, it suits lines better for applying `cmap()` as it's supposed to be applied to ends. But I'll find a workaround by applying it to all points at the ends of the cylinders I think.

Thank you very much for the help Marco!

**Reply by marcomusy** [reply]
you are welcome, you can also follow a completely different approach using implicitModeller: https://github.com/marcomusy/vedo/issues/473#issuecomment-942283048
..but then `cmap()`may become a bit more tricky.
